### PR TITLE
🚪 Property: initialize

### PR DESCRIPTION
**Pull request by Guthen**

### CODE REVIEW

**Comment by Erlite** [comment]
Fix the conflicts please

File: gamemodes/virtualroleplay/entities/weapons/vrp_keys.lua [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    timer.Simple( .2, function()
+        if IsValid( door ) then
+            door:EmitSound( "doors/door_latch3.wav" )
+        end
+    end )
+
+    --  anim
+    net.Start( "VRP:DoorAnimations" )
+        net.WriteEntity( ply )
+        net.WriteUInt( 1, 1 )
```
Comment (Erlite) [inline gamemodes/virtualroleplay/entities/weapons/vrp_keys.lua:41]: Just use WriteBit in that case

File: gamemodes/virtualroleplay/entities/weapons/vrp_keys.lua [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    net.SendPVS( ply:GetPos() )
+end
+
+local function knock( ply, sound )
+    ply:EmitSound( sound )
+
+    --  anim
+    net.Start( "VRP:DoorAnimations" )
+        net.WriteEntity( ply )
+        net.WriteUInt( 0, 1 )
```
Comment (Erlite) [inline gamemodes/virtualroleplay/entities/weapons/vrp_keys.lua:51]: Ditto

File: gamemodes/virtualroleplay/gamemode/libraries/sh_util.lua [hunk 3]
```diff
@@ -0,0 +1,25 @@
+
+local good_diff_time = .5 --  seconds
+function VRP.NetworkReceiveAsMethods( uint_bytes, methods )
```
Comment (Erlite) [inline gamemodes/virtualroleplay/gamemode/libraries/sh_util.lua:3]: One net message per function, no "shortcuts". This is a bad habit.

File: gamemodes/virtualroleplay/gamemode/modules/property/cl_property.lua [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                net.WriteUInt( line_id, VRP.PropertyMaxKeysBytes )
+            net.SendToServer()
+        end ):SetMaterial( "icon16/key_delete.png" )
+
+        menu:Open()
+    end
+end
+concommand.Add( "vrp_keys_inventory", VRP.OpenKeysInventory )
+
+net.Receive( "VRP:Property", VRP.NetworkReceiveAsMethods( 3, {
```
Comment (Erlite) [inline gamemodes/virtualroleplay/gamemode/modules/property/cl_property.lua:70]: Ditto for the different functions for the received byte.

File: gamemodes/virtualroleplay/gamemode/modules/property/sh_property.lua [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+end
+
+function PLAYER:HasPropertyKeysLimit()
+    return self.vrp_keys and #self.vrp_keys >= 2 ^ VRP.PropertyMaxKeysBytes - 1 or false
+end
+
+function PLAYER:GetLookedDoor()
+    local ent = self:GetEyeTrace().Entity
+    if not IsValid( ent ) or not ent:IsClassOwnable() then return end
+    if ent:GetPos():Distance( self:GetPos() ) > 74 then return end
```
Comment (Erlite) [inline gamemodes/virtualroleplay/gamemode/modules/property/sh_property.lua:90]: Replace with 
```lua
if ent:GetPos():DistToSqr( self:GetPos() ) > 5476 then return end
```
Reason being squared distance is more optimized.

File: gamemodes/virtualroleplay/gamemode/modules/property/sv_property.lua [hunk 6]
```diff
@@ -0,0 +1,247 @@
+util.AddNetworkString( "VRP:Property" )
+
+net.Receive( "VRP:Property", VRP.NetworkReceiveAsMethods( 3, {
```
Comment (Erlite) [inline gamemodes/virtualroleplay/gamemode/modules/property/sv_property.lua:3]: Ditto

File: gamemodes/virtualroleplay/gamemode/modules/property/sv_property.lua [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+    return owners
+end
+
+function ENTITY:SyncPropertyData( ply )
+    --  sync if owning & ownable
+    net.Start( "VRP:Property" )
+        net.WriteUInt( 1, 3 )
+        net.WriteEntity( self )
+        net.WriteTable( {
```
Comment (Erlite) [inline gamemodes/virtualroleplay/gamemode/modules/property/sv_property.lua:181]: Write table is a sin, replace if possible.

File: gamemodes/virtualroleplay/gamemode/modules/property/sv_property.lua [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+        id = id,
+        title = title,
+        --  other infos
+    }
+
+    --  sync
+    net.Start( "VRP:Property" )
+        net.WriteUInt( 2, 3 )
+        net.WriteString( "add" )
+        net.WriteTable( self.vrp_keys[#self.vrp_keys] )
```
Comment (Erlite) [inline gamemodes/virtualroleplay/gamemode/modules/property/sv_property.lua:220]: Ditto

File: gamemodes/virtualroleplay/gamemode/modules/util/sh_notification.lua [hunk 9]
```diff
@@ -16,8 +16,12 @@ if SERVER then
         end
     end
 else
-    net.Receive( "VRP:Notify", function()
-        notification.AddLegacy( net.ReadString(), net.ReadUInt( type_bytes ), net.ReadUInt( length_bytes ) )
+    function VRP.Notify( text, type, length )
```
Comment (Erlite) [inline gamemodes/virtualroleplay/gamemode/modules/util/sh_notification.lua:19]: Add sanity checks to the data received just in case.

**Review by Erlite — CHANGES_REQUESTED** [review]
Needs changes to how the networking is done, and a minor optimization tweak

**Review by Erlite — APPROVED** [review]
lgtm

**Comment by Nogitsu** [comment]
When buying a door from a double door it should buy the other.

**Review by Nogitsu — APPROVED** [review]
Everything looks good, merging.

**Comment by Guthen** [comment]
> When buying a door from a double door it should buy the other.

I don't know if I could do something about that except the thing I said in the issue #17, but it will be manually fixed.

**Pull request merged**